src/helpers/binding.ts

```diff
@@ -0,0 +1,163 @@
+const placeholderRegExp =
+  /["'`](?:(?<=")[^"\\]*(?:\\.[^"\\]*)*"|(?<=')[^'\\]*(?:\\.[^'\\]*)*'|[^`]*`)|(\?\d*|[:@$][a-zA-Z_]\w+)/gmu;
+
+function isPlainObject(obj: any): obj is Record<string, BaseType> {
+  if (typeof obj !== "object" || obj === null) return false;
+  const proto = Object.getPrototypeOf(obj);
+  return proto !== null && Object.getPrototypeOf(proto) === null;
+}
+
+function bindString(param: string, quoteEscaper = "''"): string {
+  return "'" + param.replace(/'/g, quoteEscaper) + "'";
+}
+
+function bindBoolean(param: boolean): string {
+  return Number(param).toString();
+}
+
+function bindNumber(param: number | bigint): string {
+  return param.toString();
+}
+
+function bindDate(param: Date): string {
+  return param.valueOf().toString();
+}
+
+function bindBytes(param: Uint8Array): string {
+  const hex = param.reduce((t, x) => t + x.toString(16).padStart(2, "0"), "");
+  return `X'${hex}'`;
+}
+
+function bindObject(param: any): string {
+  return bindString(JSON.stringify(param));
+}
+
+function bindNull(_param: undefined | null): string {
+  return "NULL";
+}
+
+function bindToString(param: any): string {
+  return bindString(String(param));
+}
+
+interface SQL {
+  toSQL: () => string;
+}
+
+function isSQL(param: any): param is SQL {
+  return typeof param.toSQL === "function";
+}
+
+function bindToSQL(param: SQL): string {
+  return param.toSQL();
+}
+
+function bindValue(param: BaseType): string {
+  switch (typeof param) {
+    case "bigint":
+    case "number":
+      return bindNumber(param);
+    case "boolean":
+      return bindBoolean(param);
+    case "string":
+      return bindString(param);
+    case "undefined":
+      return bindNull(param);
+    case "object":
+      if (param instanceof Date) {
```

Comment:
Yeh I considered this. But to be honest, there didn't seem like a good reason to "abuse" that pattern here. If/else are good for boolean comparisons (which have have here because of JavaScripts strange typeof/instanceof behaviors). It is true that switches are nicer for fixed comparisons (which we have here), but if you "read" the if/else branches in plain English, it is quite clear what is happening here.